### ISSUE TITLE
Strange decoding error when deserializing Avro blob

### ISSUE THREAD

**Issue by jessekempf** [GaloisInc/avro#32]
This is more of a request for debugging assistance than a bug report. I think.

I'm trying to decode RDK-B messages that use this schema: https://code.rdkcentral.com/r/plugins/gitiles/rdkb/components/opensource/ccsp/CcspLMLite/+/2f08d4d145699f31310c942199a62eb1e9f4d8b3/config/InterfaceDevicesWifi.avsc.

The actual blobs that I'm trying to decode I can't provide in this report. If it'd help, I might be able to try reproducing the issue with `QuickCheck`.

When I do a decode on a blob that's supposed to have this schema, I get:
```Error "Unexpected value when decoding for 'Double': Record (...)```, where the `(...)` is the _entire_ record, AFAICT, that’s been decoded. What’s particularly weird about this is that there’s no use of `double` _anywhere_ in the schema. I can provide the `-ddump-splices` TH output, but there's no reference to `Double` _anywhere_ in the generated code, either.

Where should I start in trying to debug this?

**Comment by AlexeyRaga** [comment]
Interesting. Can you make a branch with a unit test replicating this error?

I see only one place where that can cause this error message, which is this one:
https://github.com/GaloisInc/avro/blob/master/src/Data/Avro/FromAvro.hs#L74

Also in the error message, it says that  `Double`  doesn't match `Record`, which can have something to do either with these logical type things or with the way some double value is encoded.

Having a failing test would help to figure out what is going on, but my current hypothesis is that a record is being decoded as a value of type Double for some obscure reason.
Are you sure that the value has been encoded with the same schema?

**Comment by jessekempf** [comment]
I'm going to start on trying to make a repro.

Here's the thing, though: the code generated by TH doesn't contain the string `oub`, which means there aren't any direct references to `Double` anywhere in the generated schema. As I understand it, Avro doesn't have wire-level type identifiers -- if there is a wire identifier for a double let me know so I can look for it.

Assuming a uniform distribution over the number space of IEEE 734 double-precision floats, we have a 1 in 256 chance that the first seven bytes have their msb set, and the last byte has its msb unset, as would be required for a 64 bit varint. Otherwise the decoder should complain that it has garbled input.

I have two sample inputs which both produce the error I've mentioned.

Absent a wire identifier indicating that the following 8 bytes comprise a double, I don't see how the decoder behavior could be data-dependent. Absent a schema that defines any `Double`-valued fields, I don't see how the decoder behavior could be input-dependent.

I'm not _certain_ the value has been encoded with the same schema because I don't have access to the devices, but I'm highly confident. These avro blobs come in a message with 33 bytes of header, where the first byte is a magic number, followed by 16 bytes of schema UUID, followed by 16 bytes of schema hash. The UUID and hash match the sender source code I've been able to find. To my knowledge the schema I linked is the only version that exists.

So because I don't control the devices I can't be certain, but there's some reason to believe the schema matches.

**Comment by jessekempf** [comment]
This isn't a repro, but I've used the official Python 3 implementation of Avro to decode both the problem blobs using the schema I linked.

**Comment by jessekempf** [comment]
I serialized this python dict using the schema:
```
{'header': {}, 'cpe_id': {}, 'data': []}
```

and got the same error. This is using a version of the schema modified to eliminate the logical types use. And it's using the decoder from v0.3.0.4, SHA 65aa845.

The schema's guaranteed to match now, since blob generation is under my control. I think it pretty obvious none of these fields are secretly floating point. I think logical types have been ruled out as a possible culprit, too.

I'm going to start cutting out parts of the schema to narrow down the problem.

**Comment by jessekempf** [comment]
Right, so, I figured it out where the `Double` was coming from.

GHCi defaulting.

For some reason I had a sudden rush of shit to the brain and thought that providing an explicit schema was sufficient to indicate the type I was decoding. Nope, it doesn't work that way.

I figured this out after stripping things down so small one of the following had to be true:
1. I'm exercising code that was never tested.
2. This library doesn't actually work and is made of lies, and documentation that is also lies.
3. I'm doing something wrong.

Here's where I figured it out:
```
%> decodeWithSchema schema'InterfaceDevicesWifi <$> BSL.readFile "/tmp/avro2.avro" 
Error "Unexpected value when decoding for 'Double': Record (Record {name = \"InterfaceDevicesWifi\", namespace = Nothing, aliases = [], doc = Nothing, order = Just Ascending, fields = [Field {fldName = \"header\", fldAliases = [], fldDoc = Nothing, fldOrder = Just Ascending, fldType = String, fldDefault = Nothing}]}) (fromList [(\"header\",String \"foo\")])"
%> decodeWithSchema schema'InterfaceDevicesWifi <$> BSL.readFile "/tmp/avro2.avro" :: IO (Result InterfaceDevicesWifi)
Success (InterfaceDevicesWifi {interfaceDevicesWifiHeader = "foo"})
```
Now `/tmp/avro2.avro` literally contained the following hexadecimal contents: `06 66 6f 6f`.

Option 1 beggars belief.
Option 2 is absurd.
Option 3 is on the money.

GHCi was taking `decodeWithSchema schema'InterfaceDevicesWifi <$> BSL.readFile "/tmp/avro2.avro"` and giving it type `IO (Result Double)`.

_That_'s where `Double` was coming from.


Now that that episode of epic stupidity is out of the way, I do have confirmation that my logical types change in #31 actually works correctly. I'll add tests tomorrow.
